### PR TITLE
add e2e scripts and simplify e2e jenkins file

### PR DESCRIPTION
**Pull request by cofyc**
### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)

```
./hack/kind-cluster-build.sh 
export KUBECONFIG=$(kind get kubeconfig-path --name=kind)
make e2e
```

Migration process: https://github.com/pingcap/tidb-operator/issues/1100#issuecomment-549307011

### Does this PR introduce a user-facing change?:
 
 ```release-note

 ```

### CODE REVIEW

**Comment by cofyc** [comment]
```
2019-11-04T09:49:26.463Z] I1104 09:49:19.280759       1 actions.go:636] cleaning tidbcluster e2e/cluster1

[2019-11-04T10:00:33.059Z] I1104 10:00:22.692846       1 main.go:227] 

[2019-11-04T10:00:33.059Z] Finished.

[2019-11-04T10:00:33.059Z] panic: failed to get backup dir: failed to create pod get-backup-dir

[2019-11-04T10:00:33.059Z] 

[2019-11-04T10:00:33.059Z] goroutine 150 [running]:

[2019-11-04T10:00:33.059Z] github.com/pingcap/tidb-operator/tests/slack.NotifyAndPanic(0x1b3f1e0, 0xc000ba4e30)

[2019-11-04T10:00:33.059Z] 	/home/jenkins/agent/workspace/operator_ghpr_e2e_test_kind/go/src/github.com/pingcap/tidb-operator/tests/slack/slack.go:156 +0x271

[2019-11-04T10:00:33.059Z] github.com/pingcap/tidb-operator/tests.(*operatorActions).BackupRestoreOrDie(0xc00014b500, 0xc00000af00, 0xc00000b0e0)

[2019-11-04T10:00:33.059Z] 	/home/jenkins/agent/workspace/operator_ghpr_e2e_test_kind/go/src/github.com/pingcap/tidb-operator/tests/backup.go:201 +0xaa

[2019-11-04T10:00:33.059Z] main.main.func4(0xc000384210, 0xc00000af00, 0xc00000b0e0)

[2019-11-04T10:00:33.059Z] 	/home/jenkins/agent/workspace/operator_ghpr_e2e_test_kind/go/src/github.com/pingcap/tidb-operator/tests/cmd/e2e/main.go:166 +0x181

[2019-11-04T10:00:33.059Z] main.main.func9(0xc000384210, 0xc00022a260, 0xc00000af00, 0xc00000b0e0, 0x1bcdd40, 0xc00014b500)

[2019-11-04T10:00:33.059Z] 	/home/jenkins/agent/workspace/operator_ghpr_e2e_test_kind/go/src/github.com/pingcap/tidb-operator/tests/cmd/e2e/main.go:221 +0x6e

[2019-11-04T10:00:33.059Z] created by main.main

[2019-11-04T10:00:33.059Z] 	/home/jenkins/agent/workspace/operator_ghpr_e2e_test_kind/go/src/github.com/pingcap/tidb-operator/tests/cmd/e2e/main.go:219 +0x9b3
```

File: Makefile [hunk 1]
```diff
@@ -27,11 +28,16 @@ TEST_COVER_PACKAGES:=go list ./pkg/... | grep -vE "pkg/client" | grep -vE "pkg/t
 default: build
 
 docker-push: docker backup-docker
-	docker push "${DOCKER_REGISTRY}/pingcap/tidb-operator:latest"
-	docker push "${DOCKER_REGISTRY}/pingcap/tidb-backup-manager:latest"
+	docker push "${DOCKER_REGISTRY}/pingcap/tidb-operator:${IMAGE_TAG}"
+	docker push "${DOCKER_REGISTRY}/pingcap/tidb-backup-manager:${IMAGE_TAG}"
 
+ifeq ($(NO_BUILD),y)
```
Comment (cofyc) [inline Makefile]: NO_BUILD is added to skip building binaries because in e2e we use pre-built binaries from the previous stage

File: Makefile [hunk 2]
```diff
@@ -18,6 +18,7 @@ GOARCH := $(if $(GOARCH),$(GOARCH),amd64)
 GOENV  := GO15VENDOREXPERIMENT="1" CGO_ENABLED=0 GOOS=$(GOOS) GOARCH=$(GOARCH)
 GO     := $(GOENV) go build -trimpath
 
+IMAGE_TAG ?= latest
```
Comment (cofyc) [inline Makefile]: IMAGE_TAG is added to change image tag because in e2e we use git hash to avoid potential conflicts

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 3]
```diff
@@ -1,17 +1,6 @@
...
-	writeFile file: file, text: "${DST_E2E_FILE_CONTENT}"
-}
-
-def replace_wh(file, operatorImage) {
-	def SRC_WH_FILE_CONTENT = readFile file: file
-	def DST_WH_FILE_CONTENT = SRC_WH_FILE_CONTENT.replaceAll("image:.*", "image: ${operatorImage}")
-	writeFile file: file, text: "${DST_WH_FILE_CONTENT}"
-}
+#
+# vim: noet
```
Comment (cofyc) [inline ci/pingcap_tidb_operator_build_kind.groovy]: this file uses tabs instead of spaces, force `noexpandtab` in vim

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 4]
```diff
@@ -48,54 +37,29 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
 							sudo yum install -y mercurial
 						fi
 						hg --version
+						make check-setup
+						make check
+						make test
```
Comment (cofyc) [inline ci/pingcap_tidb_operator_build_kind.groovy]: check & verify

note that test coverage for master branch will be added in another PR (https://github.com/pingcap/tidb-operator/issues/1171)

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 5]
```diff
@@ -48,54 +37,29 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
...
 						make
-						rm -rf tests/images/test-apiserver/bin
 						make e2e-build
 						"""
 					}
 				}
-				stash excludes: "${PROJECT_DIR}/vendor/**", includes: "${PROJECT_DIR}/**", name: "tidb-operator"
+				stash excludes: "${PROJECT_DIR}/vendor/**,${PROJECT_DIR}/deploy/**", includes: "${PROJECT_DIR}/**", name: "tidb-operator"
```
Comment (cofyc) [inline ci/pingcap_tidb_operator_build_kind.groovy]: terraform modules is very large, so we should skip them here to speed up this process

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 6]
```diff
@@ -135,78 +99,8 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
...
-						elif [[ \$status == Failed ]]
-						then
-							exit 1
-						fi
-						execType=1
-					done
-					trap - INT TERM ERR
-					rm -f /k8s-locks/\$clusterName
+                    export KUBECONFIG=`/root/go/bin/kind get kubeconfig-path --name="\$clusterName"`
+                    DOCKER_REGISTRY=\${clusters[\$clusterName]} IMAGE_TAG=${GITHASH} SKIP_BUILD=y make e2e
```
Comment (cofyc) [inline ci/pingcap_tidb_operator_build_kind.groovy]: move all image build and e2e running logic into hack script

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 7]
```diff
@@ -258,3 +152,5 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
 	}
 }
 return this
+
+// vim: noet
```
Comment (cofyc) [inline ci/pingcap_tidb_operator_build_kind.groovy]: this file uses tabs, force vim `noexpandtab`.

File: deploy/modules/gcp/tidb-operator/versions.tf [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+terraform {
+  required_version = ">= 0.12"
+  required_providers {
+    google      = "~> 2.16"
+    google-beta = "~> 2.16"
+    external    = "~> 1.2"
+    helm        = "~> 0.10"
+    null        = "~> 2.1"
+  }
+}
```
Comment (cofyc) [inline deploy/modules/gcp/tidb-operator/versions.tf]: `deploy/modules` do not contain version-lock file, so they will install the latest provider and may fail on `validate` command
xref: https://github.com/pingcap/tidb-operator/issues/1172

File: hack/run-e2e.sh [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+kubectl delete -f tests/manifests/e2e/e2e.yaml --ignore-not-found
+kubectl wait --for=delete -n ${NS} pod/tidb-operator-e2e || true
+# Create sa first and wait for the controller to create the secret for this sa
+# This avoids `No API token found for service account " tidb-operator-e2e"`
+# TODO better way to work around this issue
+kubectl -n ${NS} create sa tidb-operator-e2e
+kubectl -n ${NS} apply -f $tmpfile
+kubectl -n ${NS} wait --for=condition=Ready pod/tidb-operator-e2e
+kubectl -n ${NS} logs -f tidb-operator-e2e
```
Comment (cofyc) [inline hack/run-e2e.sh]: should check the finish status of e2e pod, I'm working on it.

Comment (cofyc) [reply hack/run-e2e.sh]: done

File: hack/run-e2e.sh [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+echo "info: checking e2e result"
+phrase=$(kubectl -n ${NS} get pods tidb-operator-e2e -ojsonpath='{.status.phase}')
```
Comment (tennix) [inline hack/run-e2e.sh]:
```suggestion
phase=$(kubectl -n ${NS} get pods tidb-operator-e2e -ojsonpath='{.status.phase}')
```

File: hack/e2e.sh [hunk 11]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Comment (tennix) [inline hack/e2e.sh]:
```suggestion
#!/usr/bin/env bash
```

File: hack/run-e2e.sh [hunk 12]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
```
Comment (tennix) [inline hack/run-e2e.sh]:
```suggestion
#!/usr/bin/env bash
```

File: hack/run-e2e.sh [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+echo "TIDB_OPERATOR_IMAGE: $TIDB_OPERATOR_IMAGE"
+echo "E2E_IMAGE: $E2E_IMAGE"
+echo "TEST_APISERVER_IMAGE: $TEST_APISERVER_IMAGE"
+echo "KUBECONFIG: $KUBECONFIG"
+
+NS=tidb-operator-e2e
+
+# copy and modify to avoid local changes
+# TODO find a better way
```
Comment (tennix) [inline hack/run-e2e.sh]: Is it possible to use `cat tests/manifests/e2e/e2e.yaml | sed 'xxx' | kubectl apply -f -`?

Comment (cofyc) [reply hack/run-e2e.sh]: done

**Comment by tennix** [comment]
@onlymellb PTAL

File: hack/run-e2e.sh [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+# This avoids `No API token found for service account " tidb-operator-e2e"`
+# TODO better way to work around this issue
+kubectl -n ${NS} create sa tidb-operator-e2e
+# copy and modify to avoid local changes
+sed "s#image: localhost:5000/pingcap/tidb-operator-e2e:latest#image: $E2E_IMAGE#g
+s#=pingcap/tidb-operator:latest#=${TIDB_OPERATOR_IMAGE}#g
+s#=pingcap/test-apiserver:latest#=${TEST_APISERVER_IMAGE}#g
+" tests/manifests/e2e/e2e.yaml | kubectl -n ${NS} apply -f -
+kubectl -n ${NS} wait --for=condition=Ready pod/tidb-operator-e2e
+kubectl -n ${NS} logs -f tidb-operator-e2e
```
Comment (onlymellb) [inline hack/run-e2e.sh]: Here we need to loop to output the e2e pod's log. According to the previous experience, the connection to get the log will occasionally be disconnected. We need to re-export the e2e pod's log to the standard output after the connection is broken. So we need to keep the logic in the previous groovy script.

Comment (cofyc) [reply hack/run-e2e.sh]: done

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 15]
```diff
@@ -48,54 +37,29 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
...
-					cp -r charts/tidb-cluster tests/images/e2e
-					cp -r charts/tidb-backup tests/images/e2e
-					cp -r manifests tests/images/e2e
-					docker build -t ${E2E_IMAGE} tests/images/e2e
-					if [ -f tests/images/test-apiserver/bin/tidb-apiserver ]; then docker build -t ${APISERVER_IMAGE} tests/images/test-apiserver; fi
-					"""
-				}
-
 				stage('start run operator e2e test'){
 					def operator_yaml = "tests/manifests/e2e/e2e.yaml"
```
Comment (onlymellb) [inline ci/pingcap_tidb_operator_build_kind.groovy]: Remove this line?

File: ci/pingcap_tidb_operator_build_kind.groovy [hunk 16]
```diff
@@ -48,54 +37,29 @@ def call(BUILD_BRANCH, CREDENTIALS_ID) {
...
-					replace(operator_yaml, IMAGE_TAG, E2E_IMAGE, APISERVER_IMAGE)
 					ansiColor('xterm') {
 					def ns = "tidb-operator-e2e"
```
Comment (onlymellb) [inline ci/pingcap_tidb_operator_build_kind.groovy]: ditto

Comment (cofyc) [reply ci/pingcap_tidb_operator_build_kind.groovy]: done

File: ci/pingcap_tidb_operator_build_kind.groovy (hunk 15 above)
Comment (cofyc) [reply ci/pingcap_tidb_operator_build_kind.groovy]: done

File: hack/run-e2e.sh (hunk 14 above)
Comment (cofyc) [reply hack/run-e2e.sh]: done

File: hack/run-e2e.sh [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+    if [[ "$phase" == "Succeeded" ]]; then
+        echo "info: e2e succeeded"
+        exit 0
+    elif [[ "$phase" == "Failed" ]]; then
+        echo "error: e2e failed, phase: $phase"
+        exit 1
+    else
+        continue
+    fi
+done
```
Comment (cofyc) [inline hack/run-e2e.sh]: I misunderstood this part, will update it soon.

**Comment by cofyc** [comment]
@tennix @onlymellb PTAL again.

**Comment by cofyc** [comment]
made a minor update on `kubectl logs` command
@tennix @onlymellb PTAL

File: tests/actions.go [hunk 18]
```diff
@@ -472,7 +472,11 @@ func (oa *operatorActions) DeployOperator(info *OperatorConfig) error {
 	}
 
 	// deploy statefulset webhook and configuration to hijack update statefulset opeartion
-	cmd = fmt.Sprintf(`sed 's/apiVersions: \["v1beta1"\]/apiVersions: ["v1", "v1beta1"]/' %s/webhook.yaml | kubectl apply -f -`, oa.manifestPath(info.Tag))
+	cmd = fmt.Sprintf(`
+sed 's/apiVersions: \["v1beta1"\]/apiVersions: ["v1", "v1beta1"]/
+s#image:.*#image: %s#g
+' %s/webhook.yaml | kubectl apply -f -
+`, info.Image, oa.manifestPath(info.Tag))
```
Comment (cofyc) [inline tests/actions.go]: webhook.yaml is packed into the e2e image, we should update its image here
when we migrate the webhook deployment into the helm chart, this can be simplified
